### ISSUE TITLE
Element-wise tensor ops do not broadcast implicitly

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#918]
## Summary

`CpuEngine`'s element-wise tensor operations require **exactly matching shapes** and throw otherwise. They do not broadcast. A parallel `TensorBroadcast*` family exists to cover the gap, so every caller has to know in advance whether its operands will line up and pick a different method accordingly.

NumPy and PyTorch both broadcast implicitly. This is the single largest ergonomic gap between this engine and what people expect from a tensor library.

## Evidence

```
CpuEngine.cs:3004   TensorAdd       if (!ShapesMatch(a._shape, b._shape)) throw
CpuEngine.cs:3173   TensorAddInPlace
CpuEngine.cs:5315   ...
CpuEngine.cs:5603   ...
        "Tensor shapes must match. Got [4, 3] and [4, 1]."
```

The broadcasting-capable alternatives:

```
IEngine.cs:1979   TensorBroadcastTo(input, targetShape)
IEngine.cs:2080   TensorBroadcastAdd(a, b)
IEngine.cs:2267   TensorBroadcastSubtract(a, b)
IEngine.cs:2283   TensorBroadcastDivide(a, b)
IEngine.cs:2299   TensorBroadcastMultiply(a, b)
```

## How it surfaced

Making the self-supervised objectives differentiable in AiDotNet (InfoNCE, NT-Xent, BYOL, DINO, Barlow Twins, MAE). Every one of them reduces along an axis and then combines the reduced tensor back with the full one — L2 row normalization divides `[rows, cols]` by `[rows, 1]`, stable log-softmax subtracts a `[rows, 1]` row max, Barlow Twins standardizes `[batch, dim]` against a `[1, dim]` mean.

Written the way the math reads, all of it threw. It only worked after switching each site to the `TensorBroadcast*` variant. Nothing about the call site suggests which family to reach for; you find out at runtime.

## The trade-off, stated honestly

**For implicit broadcasting**

- Matches NumPy, PyTorch, JAX and TensorFlow. Least-surprise for anyone arriving from those.
- Removes a whole class of avoidable runtime failures in numerically ordinary code.
- Collapses two parallel op families into one, so there is no longer a wrong-but-compiling choice.
- Reduction-then-recombine is the single most common shape pattern in ML code, and it is exactly the pattern that currently fails.

**Against**

- It converts loud failures into silent successes. `[4,1] + [4,3]` throws today; afterwards it quietly yields `[4,3]`. A transposed operand or an off-by-one axis stops announcing itself.
- That is the same failure mode as the shape bugs that dominated ooples/AiDotNet#1789: shapes disagreeing quietly and surfacing far from the cause. Several models there declared one shape and computed another for exactly this reason.
- The strictness may be load-bearing in places that currently rely on the throw as a guard.

## Options

1. **Implicit broadcasting on the element-wise ops**, `TensorBroadcast*` kept as aliases. Matches the wider ecosystem; accepts the silent-success risk.
2. **Implicit broadcasting behind an engine-level opt-in flag**, defaulting to strict. Preserves current guarantees while letting new code opt in; the cost is two behaviours to reason about.
3. **Keep strict, improve the failure.** Have the exception name the broadcasting alternative and the shapes that would have worked — turning a dead end into a signpost. Cheapest, and keeps loud failures loud, but leaves the ergonomic gap.

## Recommendation

Option 1 if the goal is to match what users expect from a tensor library, which is the stated direction. If we take it, it should land with tests that pin the intended broadcast semantics per rank combination, so "it broadcast something" never gets mistaken for "it broadcast correctly" — and ideally with a shape-assertion helper available for the call sites that genuinely want strictness.

Worth deciding deliberately rather than by default: the argument against is not theoretical, it is the exact bug class that produced 850 failing tests in the consuming repo.
